hashicorp/dev-portal
# feat: variant rewrite state

## 🔗 Relevant links



- [Preview link](https://dev-portal-git-ksvariant-rewrite-state-hashicorp.vercel.app/) 🔎
- [Asana task](https://app.asana.com/0/1204333057896641/1204621999504621) 🎟️

## 🗒️ What

Builds a map of rewrites that are generated to a static file at build time. These are imported and read in middleware where the path rewriting logic occurs, depending on the param and cookie. Data structure for the generated file looks like the image below:

<img width="776" alt="Screenshot 2023-05-23 at 4 11 17 PM" src="https://github.com/hashicorp/dev-portal/assets/36613477/0dfe9ff3-ed79-4b8b-b613-ff205cc2beb2">


## 🤷 Why



## 🛠️ How



## 📸 Design Screenshots



## 🧪 Testing

- Visit these test routes: [kubernetes-api-gateway](https://dev-portal-git-ksvariant-rewrite-state-hashicorp.vercel.app/consul/tutorials/kubernetes/kubernetes-api-gateway), [get-started-nomad-cloud](https://dev-portal-git-ksvariant-rewrite-state-hashicorp.vercel.app/waypoint/tutorials/get-started-nomad-cloud/get-started-nomad-cloud) 
- Select a non-default variant option (2nd option), or verify that you have cookies set for the tutorial variant option that you're on, the route will change to show search params when you select the option. 
- Update your url to [the base route ](https://dev-portal-git-ksvariant-rewrite-state-hashicorp.vercel.app/consul/tutorials/kubernetes/kubernetes-api-gateway) without the query param and your selected variant option should persist, while the url shouldn't visibly change. 
- Navigate to the route via the sidebar, validate that the correct variant option renders even on the base URL now. 
- Navigate to a [query param path t](https://dev-portal-git-ksvariant-rewrite-state-hashicorp.vercel.app/consul/tutorials/kubernetes/kubernetes-api-gateway?variants=consul-deploy%3Ahcp)hat isn't your stored preference, that variant should be active since the query param takes precedent. 

## 💭 Anything else?

There's a [known issue](https://app.asana.com/0/1204333057896641/1204651148003825) with the sidebar, where the active list item isn't calculating properly when directly navigating to query param routes.
